client/src/components/WebAuthn/Verify.js

```diff
@@ -2,87 +2,198 @@
+   *
+   * @returns {HTMLElement}
+   */
+  renderStatus() {
+    const { ss: { i18n } } = window;
+    const { failure, success } = this.state;
 
-  render() {
+    if (success) {
+      return (
+        <div className="mfa-verification-container__status status-message--success">
+          <span className="status-message__icon"><CircleTick size="32px" /></span>
+          <span className="status-message__description">
+            {i18n._t('MFAWebAuthnVerify.SUCCESS', fallbacks['MFAWebAuthnVerify.SUCCESS'])}
+          </span>
+        </div>
+      );
+    }
+
+    if (failure) {
+        return (
+          <div className="mfa-verification-container__status status-message--failure">
+            <span className="status-message__icon"><CircleWarning size="32px" /></span>
+            <span className="status-message__description">
+              {i18n._t('MFAWebAuthnVerify.FAILURE', fallbacks['MFAWebAuthnVerify.FAILURE'])}
+            </span>
+          </div>
+        );
+    }
+
+    return (
+      <div className="mfa-verification-container__status status-message--loading">
+        <LoadingIndicator size="3em" />
+        <span className="status-message__description">
+          {i18n._t('MFAWebAuthnVerify.WAITING', fallbacks['MFAWebAuthnVerify.WAITING'])}
+        </span>
+      </div>
+    );
+  }
+
+  /**
+   * Render the icon related to this method
+   *
+   * @returns {HTMLElement}
+   */
+  renderThumbnail() {
+    const { method } = this.props;
+
+    return (
+      <div className="mfa-verification-container__thumbnail">
+        <img
+          src={method.thumbnail}
+          alt={method.name}
+        />
+      </div>
+    );
+  }
+
+  /**
+   * Render available actions based on current state of registration flow
+   *
+   * @return {HTMLElement}
+   */
+  renderActions() {
+    const { ss: { i18n } } = window;
     const { moreOptionsControl } = this.props;
+    const { failure, success } = this.state;
+
+    if (success) {
+      return <div className="mfa-verification-container__actions" />;
+    }
+
+    const retryName = i18n._t('MFAWebAuthnVerify.RETRY', fallbacks['MFAWebAuthnVerify.RETRY']);
+    const retryButton = (
+      <button
+        key={retryName}
+        className="btn btn-primary"
+        disabled={false}
+        onClick={this.handleRetry}
+        type="button"
+      >
+        {retryName}
+      </button>
+    );
 
     return (
-      <form className="mfa-verify-web-authn__container">
-        <div className="mfa-verify-web-authn__content">
-          {this.renderDescription()}
-        </div>
-        <button onClick={this.handleStartAuth}>Do it</button>
+      <div className="mfa-verification-container__actions">
+        {failure ? retryButton : null}
         {moreOptionsControl}
-        <div className="mfa-verify-web-authn__icon" />
-      </form>
+      </div>
+    );
+  }
+
+  render() {
+    const { failure, success } = this.state;
+
+    if (!(failure || success)) {
+      this.startAuth();
```

Comment:
You shouldn't really put state changes in the render method, could you move this perhaps to the `componentDidMount()` method or something like that?

If I'm reading it correctly, this is saying "if there's no failure or success marker then show the webauthn verification dialog" - if that's the case then you can do that in `componentDidUpdate()` I think

Reply:
No, it waits forever.

Reply:
What does, sorry? If you must put this here, can you change its name to `renderBrowserAuth()` or something so it's clearer what it's doing?

Reply:
If I begin the webauthn API requests within `componentDidUpdate` instead of in `render` then it renders the "Loading" screen, and doesn't actually start any authentication.

You're right in that it's not actually rendering anything (in react terms), so I'm not sure `renderBrowserAuth` would be quite right either.

In that the `startAuth` method starts the authentication API process, I think it is aptly named. It doesn't actually render anything in the React sense, so I can see the aversion to it being in the `render` function. However in lack of a known more defined place to put such a call, I don't think I agree with the level of objection; to me it currently reads "when rendering, first start the authentication process before drawing elements" (e.g. it might be known as `onBeforeRender` in another world), and to put it in another method for the sake of it would bit a little superfluous.